ospec/tests/test-ospec.js

```diff
@@ -178,104 +93,178 @@ o.spec("ospec", function() {
+				timeout(40)
+
+				setTimeout(done, 50)
+			})
+
+			clone.run(function(results){
+				o(results.every(function(report){return report.pass === true}))
 
 				done()
 			})
 		})
 	})
 
-	o.spec("stack trace cleaner", function() {
-		o("handles line breaks", function() {
-			try {
-				throw new Error("line\nbreak")
-			} catch(error) {
-				var trace = o.cleanStackTrace(error)
-				o(trace).notEquals("break")
-				o(trace.includes("test-ospec.js")).equals(true)
-			}
-		})
-	})
+	// Predicate test passing on clone results
+	o.spec("reporting", function () {
+		var clone = o.new()
 
-	o.spec("async promise", function() {
-		var a = 0, b = 0
-
-		function wrapPromise(fn) {
-			return new Promise((resolve, reject) => {
-				callAsync(() => {
-					try {
-						fn()
-						resolve()
-					} catch(e) {
-						reject(e)
-					}
-				})
+		clone.spec("clone", function () {
+			clone("fail", function () {
+				clone(true).equals(false)
 			})
-		}
 
-		o.before(function() {
-			return wrapPromise(() => {
-				a = 1
+			clone("pass", function () {
+				clone(true).equals(true)
 			})
 		})
 
-		o.after(function() {
-			return wrapPromise(function() {
-				a = 0
-			})
-		})
+		o("reports per instance", function () {
+			clone.run(function (results) {
+				o(typeof results).equals("object")
+				o("length" in results).equals(true)
+				o(results.length).equals(2)("Two results")
 
-		o.beforeEach(function() {
-			return wrapPromise(function() {
-				b = 1
+				o("error" in results[0] && "pass" in results[0]).equals(true)("error and pass keys present in failing result")
+				o(!("error" in results[1]) && "pass" in results[1]).equals(true)("only pass key present in passing result")
+				o(results[0].pass).equals(false)("Test meant to fail has failed")
+				o(results[1].pass).equals(true)("Test meant to pass has passed")
 			})
 		})
-		o.afterEach(function() {
-			return wrapPromise(function() {
-				b = 0
-			})
+
+		o("o.report() returns the number of failures", function () {
+			var log = console.log, error = console.error
+			console.log = o.spy()
+			console.error = o.spy()
+
+			function makeError(msg) { try { throw msg ? new Error(msg) : new Error } catch (e) { return e } }
+			try {
+				var errCount = o.report([{ pass: true }, { pass: true }])
+
+				o(errCount).equals(0)
+				o(console.log.callCount).equals(1)
+				o(console.error.callCount).equals(0)
+
+				errCount = o.report([
+					{ pass: false, error: makeError("hey"), message: "hey" }
+				])
+
+				o(errCount).equals(1)
+				o(console.log.callCount).equals(2)
+				o(console.error.callCount).equals(1)
+
+				errCount = o.report([
+					{ pass: false, error: makeError("hey"), message: "hey" },
+					{ pass: true },
+					{ pass: false, error: makeError("ho"), message: "ho" }
+				])
+
+				o(errCount).equals(2)
+				o(console.log.callCount).equals(3)
+				o(console.error.callCount).equals(3)
+			} catch (e) {
+				o(1).equals(0)("Error while testing the reporter")
```

Comment:
`o("Error while testing the reporter").equals(null)`?

Reply:
I've tried those it and the reporting just jars with my reading mode. I far prefer assertions that are either transparently meaningless or totally salient - the failure case for this would drive me up the wall. Besides, I think the ospec test suite should be a guide for how to write ospec - the least we can do is dog food the API affordances.

Reply:
Good points, agreed!

Reply:
IMO ospec needs an explicit "this should not run" method – `o.fail(optionalMessage)` for example

Reply:
I see where you're coming from, but I prefer your earlier idea for `o.fail` that did a drastically different thing and signalled code that was *expected* to fail. Isn't the simplest thing in this case just to `throw`?